src/pages/ProfilePage.tsx

```diff
@@ -89,31 +68,93 @@ export default function ProfilePage() {
         }
 
         const newProfileData: ProfileData = {
-          username: 'testuser',
+          username: data.username,
           email: data.email,
           name: data.name,
-          school: 'University of Maryland',
+          school: data.school,
           numAttempts: data.problemsAttempted,
           pieChartData: newPieChartData,
         }
 
+        setOldProfileData(newProfileData)
         setProfileData(newProfileData)
+        setIsLoadingPage(false)
       })
       .catch((error: Error) => {
-        console.log(
-          'Error fetching data for user ' + user.id + ': ' + error.message,
-        )
+        console.error(error)
       })
   }, [])
 
-  const onTextFieldChange = (key, newString) => {
-    setProfileData((oldData: ProfileData) => {
-      return { ...oldData, [key]: newString }
+  async function submitEdit(): Promise<void> {
+    interface ProfileEditQuery {
+      uid: string
+      username?: string
+      email?: string
+      name?: string
+      school?: string
+    }
+
+    // compare old data to new data, and only submit the fields that were changed.
+    const sendToAPI: ProfileEditQuery = { uid: user.id }
+    if (oldProfileData.username !== profileData.username) {
+      sendToAPI.username = profileData.username
+    }
+    if (oldProfileData.email !== profileData.email) {
+      sendToAPI.email = profileData.email
+    }
+    if (oldProfileData.name !== profileData.name) {
+      sendToAPI.name = profileData.name
+    }
+    if (oldProfileData.school !== profileData.school) {
+      sendToAPI.school = profileData.school
+    }
+
+    setIsWaiting(true)
+    fetch(buildPath('/updateUserData'), {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(sendToAPI),
     })
+      .then((res: Response) => {
+        if (res.ok) {
+          return res.json()
+        }
+        throw Error(res.statusText)
+      })
+      .then((data) => {
+        setIsWaiting(false)
+
+        let editWasSuccessful: boolean = true
+        for (const key of Object.keys(data)) {
+          if (data[key] !== 'Not Updated' && data[key] !== 'Success') {
+            editWasSuccessful = false
+
+            // Revert this item (key) back to what it was before editing.
+            setProfileData((prevData: ProfileData) => {
+              return { ...prevData, [key]: oldProfileData[key] }
+            })
+          }
+        }
+
+        setUsernameStatus(data.username)
+        setEmailStatus(data.email)
+
+        if (editWasSuccessful) {
+          setIsEditing(false)
+        }
```

Comment:
This might be an important one. If successful, we probably also want to update oldProfileData to the new data, right? Otherwise, if you edit again and the API call fails, it will reset to the very old profile data before the successful call.